File: components/data_sharing/public/android/java/src/org/chromium/components/data_sharing/ServiceStatus.java
// Copyright 2024 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.components.data_sharing;

import org.jni_zero.CalledByNative;
import org.jni_zero.JNINamespace;

/** Information about a member of a group. */
@JNINamespace("data_sharing")
public class ServiceStatus {
    public @SigninStatus int signinStatus;
    public @SyncStatus int syncStatus;
    public @CollaborationStatus int collaborationStatus;

    /**
     * Constructor for a {@link ServiceStatus} object.
     *
     * @param signinStatus The signin status of the service.
     * @param syncStatus The sync status of the service.
     * @param collaborationStatus The collaboration status of the service.
     */
    public ServiceStatus(
            @SigninStatus int signinStatus,
            @SyncStatus int syncStatus,
            @CollaborationStatus int collaborationStatus) {
        this.signinStatus = signinStatus;
        this.syncStatus = syncStatus;
        this.collaborationStatus = collaborationStatus;
    }

    // Helper functions for checking DataSharingService's status.
    // Please keep the logic consistent with //components/data_sharing/public/service_status.h.

    /**
     * @return Whether the user is allowed to join a Data Sharing collaboration.
     */
    public boolean isAllowedToJoin() {
        switch (collaborationStatus) {
            case CollaborationStatus.DISABLED:
            case CollaborationStatus.DISABLED_FOR_POLICY:
                return false;
            case CollaborationStatus.ALLOWED_TO_JOIN:
            case CollaborationStatus.ENABLED_JOIN_ONLY:
            case CollaborationStatus.ENABLED_CREATE_AND_JOIN:
                return true;
        }
        return false;
    }

    /**
     * @return Whether the user is allowed to create a Data Sharing collaboration.
     */
    public boolean isAllowedToCreate() {
        switch (collaborationStatus) {
            case CollaborationStatus.DISABLED:
            case CollaborationStatus.DISABLED_FOR_POLICY:
            case CollaborationStatus.ALLOWED_TO_JOIN:
            case CollaborationStatus.ENABLED_JOIN_ONLY:
                return false;
            case CollaborationStatus.ENABLED_CREATE_AND_JOIN:
                return true;
        }
        return false;
    }

    @CalledByNative
    private static ServiceStatus createServiceStatus(
            int signinStatus, int syncStatus, int collaborationStatus) {
        return new ServiceStatus(signinStatus, syncStatus, collaborationStatus);
    }
}

File: components/collaboration/public/service_status.cc
// Copyright 2024 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "components/collaboration/public/service_status.h"

namespace collaboration {

// LINT.IfChange(ServiceStatus)
bool ServiceStatus::IsAllowedToJoin() {
  // Please keep logic consistent with
  // //components/collaboration/public/android/java/src/org/chromium/components/collaboration/ServiceStatus.java.
  switch (collaboration_status) {
    case CollaborationStatus::kDisabled:
    case CollaborationStatus::kDisabledForPolicy:
      return false;
    case CollaborationStatus::kAllowedToJoin:
    case CollaborationStatus::kEnabledJoinOnly:
    case CollaborationStatus::kEnabledCreateAndJoin:
      return true;
  }
}

bool ServiceStatus::IsAllowedToCreate() {
  // Please keep logic consistent with
  // //components/collaboration/public/android/java/src/org/chromium/components/collaboration/ServiceStatus.java.
  switch (collaboration_status) {
    case CollaborationStatus::kDisabled:
    case CollaborationStatus::kDisabledForPolicy:
    case CollaborationStatus::kAllowedToJoin:
    case CollaborationStatus::kEnabledJoinOnly:
      return false;
    case CollaborationStatus::kEnabledCreateAndJoin:
      return true;
  }
}
// LINT.ThenChange(//components/collaboration/public/android/java/src/org/chromium/components/collaboration/ServiceStatus.java)

}  // namespace collaboration
